pkg/virtual/workspaces/registry/rest.go

```diff
@@ -0,0 +1,482 @@
+	}
+
+	prettyName := workspace.Name
+	var createdWorkspace *tenancyv1alpha1.Workspace
+	var err error
+	var nameSuffix string
+	i := 0
+	for i < 10 {
+		if i > 0 {
+			nameSuffix = fmt.Sprintf("-%d", i)
+			workspace.Name = fmt.Sprintf("%s-%s", prettyName, nameSuffix)
+		}
+		createdWorkspace, err = s.workspaceClient.Create(ctx, workspace, metav1.CreateOptions{})
+		if err == nil {
+			break
+		}
+		if !kerrors.IsAlreadyExists(err) {
+			return nil, err
+		}
+		i++
+	}
+
+	adminClusterRole.Rules[0].ResourceNames = []string{createdWorkspace.Name}
+	adminClusterRole.Labels[InternalNameLabel] = createdWorkspace.Name
+	if _, err := s.rbacClient.ClusterRoles().Update(ctx, &adminClusterRole, metav1.UpdateOptions{}); err != nil {
+		_ = s.rbacClient.ClusterRoles().Delete(ctx, adminClusterRole.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		_ = s.rbacClient.ClusterRoles().Delete(ctx, listerClusterRole.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		_, _, _ = s.Delete(ctx, createdWorkspace.Name, nil, &metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		if kerrors.IsConflict(err) {
+			return nil, kerrors.NewConflict(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+		}
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	listerClusterRole.Rules[0].ResourceNames = []string{createdWorkspace.Name}
+	listerClusterRole.Labels[InternalNameLabel] = createdWorkspace.Name
+	if _, err := s.rbacClient.ClusterRoles().Update(ctx, &listerClusterRole, metav1.UpdateOptions{}); err != nil {
+		_ = s.rbacClient.ClusterRoles().Delete(ctx, adminClusterRole.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		_ = s.rbacClient.ClusterRoles().Delete(ctx, listerClusterRole.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		_, _, _ = s.Delete(ctx, createdWorkspace.Name, nil, &metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		if kerrors.IsConflict(err) {
+			return nil, kerrors.NewConflict(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+		}
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	clusterRoleBinding.Labels[InternalNameLabel] = createdWorkspace.Name
+	clusterRoleBinding.Labels[PrettyNameLabel] = prettyName
+	if _, err := s.rbacClient.ClusterRoleBindings().Update(ctx, &clusterRoleBinding, metav1.UpdateOptions{}); err != nil {
+		var zero int64
+		_ = s.rbacClient.ClusterRoleBindings().Delete(ctx, clusterRoleBinding.Name, metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		_, _, _ = s.Delete(ctx, createdWorkspace.Name, nil, &metav1.DeleteOptions{GracePeriodSeconds: &zero})
+		if kerrors.IsConflict(err) {
+			return nil, kerrors.NewConflict(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+		}
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspaces"), workspace.Name, err)
+	}
+
+	createdWorkspace.Name = prettyName
+	return createdWorkspace, nil
+}
```

Comment:
please add some go docs, a line above every block, saying what it does and why.

Reply:
Added detailed GO doc inside the function to explain why we do things this way.